### PR TITLE
Fix #294 - blacklist for slug validation

### LINKED ISSUE #294: Create a blacklist for slug validation
**Hey!** If you ever wanted to contribute to an Open Source project, but didn't know how, you just found a perfect place to start. We try to write this issue as instructive as possible, but as we're humans we almost certainly forgot to write something important in here. If you want to try to tackle this issue and get stuck at _any_ point, please do not hesitate and ask us, either here on GitHub or via email at support@pretix.eu. We'll be very happy to help! 

We think this issue is very suitable to people who are just starting with Python or Django development or who never contributed to Open Source before. We'd like to ask people with wide experience in these areas to leave this issue to beginners. Thanks! (If you want something more advanced, just drop us a line, we'll find something :wink:)

## How to get started

First of all, you need pretix running locally on your machine. We have a guide on the developer installation [over here](https://docs.pretix.eu/en/latest/development/setup.html). Play around with it and get a feeling for what pretix is and how it works :) Try to find the places in the interface mentioned in this issue where something needs to change. Maybe try to find the same places in the code? Just searching for characteristic words can be a powerful tool to find your way around.

## What this is about

Every event organized with pretix has a "slug", a short version of its name that we use for URLs. The same goes for organizers, so we often build URLs like ``https://pretix.eu/<organizer-slug>/<event-slug>``. This makes for very nice URLs, but it has the danger of conflicts with internal URLs of pretix, for example the pretix backend is at ``/control/``. What about an organizer with the slug ``control``? This will lead to confusion and should be forbidden.

## How to do this

To avoid URL conflicts, we must validate that slugs (the short names for events and organizers that we use in URLs) do not have the following values:

Event slugs:
* ``download``
* ``healthcheck``
* ``locale``
* ``control``
* ``redirect``
* ``jsi18n``
* ``metrics`` (#271)
* ``_global``
* ``__debug__``´

Organizer slugs:
* ``download``
* ``healthcheck``
* ``locale``
* ``control``
* ``pretixdroid``
* ``redirect``
* ``jsi18n``
* ``metrics`` (#271)
* ``_global``
* ``__debug__``
* ``about``

We can do this by writing a custom [validator](https://docs.djangoproject.com/en/1.10/ref/validators/) and then add this to the slug fields in our ``Event`` and ``Organizer`` models.

## I want to give it a try!

You're still reading, nice! If you want to try working on this, please comment below, so everyone else knows this is being looked at and can search for a different place to contribute. 

First, [fork](https://github.com/pretix/pretix/fork) the repository to your own GitHub account. Then clone it locally (or [set it as your remote](https://help.github.com/articles/configuring-a-remote-for-a-fork/) if you already cloned it earlier) and prepare your pull requests. There are a number of tutorials on the Pull Request workflow online, for example this [short version](https://guides.github.com/introduction/flow/) by GitHub itself and this [more in-depth tutorial](https://www.digitalocean.com/community/tutorials/how-to-create-a-pull-request-on-github). We're excited to see your pull request! 

And don't forget: If you get stuck, don't hestitate to ask. We bound ourselves to a [Code of Conduct](https://docs.pretix.eu/en/latest/development/contribution/codeofconduct.html) and we're committed to providing a welcoming experience and giving constructive, friendly and useful feedback on your contributions.

### PR DESCRIPTION
**Pull request by chrisdamba**
Not sure about a couple of things - the validation message correctness and the most appropriate place to place the unit tests.

### CODE REVIEW

File: src/pretix/base/validators.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class BlacklistValidator:
+
+    blacklist = []
+
+    def __call__(self, value):
+        # Validation logic
+        if value in self.blacklist:
+            message = _("This slug has an invalid value.")
```
Comment (rixx) [inline src/pretix/base/validators.py]: I think having the failing slug in the message would be even better (appended, for example).

Comment (raphaelm) [inline src/pretix/base/validators.py]: If you want to change this, see also https://docs.djangoproject.com/en/1.10/ref/forms/validation/#raising-validationerror about how to use parameters here

**Comment by raphaelm** [comment]
I think the location of the tests is okay

File: src/pretix/base/validators.py (hunk 1 above)
Comment (chrisdamba) [inline src/pretix/base/validators.py]: I think that's a great idea too

File: src/pretix/base/validators.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    blacklist = []
+
+    def __call__(self, value):
+        # Validation logic
+        if value in self.blacklist:
+            raise ValidationError(
+                _('This slug has an invalid value: %(value)s.'),
+                code='invalid',
+                params={'value': '42'},
```
Comment (raphaelm) [inline src/pretix/base/validators.py]: You probably mean ``value`` here instead of ``'42'`` :wink:

**Comment by raphaelm** [comment]
``EventTest.test_slug_validation`` fails on travis. Does it work for you locally?

File: src/pretix/base/validators.py (hunk 2 above)
Comment (chrisdamba) [inline src/pretix/base/validators.py]: *facepalms* You are so right, thanks for the correction :)

**Comment by chrisdamba** [comment]
Was running fine locally after correcting the datetime typo. May need to check how I'm creating the event object in that failing test else I'm not sure why it's not raising the ValidationError on Travis CI test

**Comment by raphaelm** [comment]
It fails for me locally as well. The solution is to call ``full_clean()`` in the test on the model, not ``clean()``. https://docs.djangoproject.com/en/1.10/ref/models/instances/#django.db.models.Model.full_clean

**Comment by raphaelm** [comment]
Looks good, thank you very much!

**Pull request merged**